Source/WebCore/page/LocalDOMWindow.cpp

```diff
@@ -1611,6 +1619,23 @@ void LocalDOMWindow::consumeLastActivationIfNecessary()
         m_lastActivationTimestamp = -MonotonicTime::infinity();
 }
 
+// https://html.spec.whatwg.org/multipage/interaction.html#consume-history-action-user-activation
+bool LocalDOMWindow::consumeHistoryActionUserActivation()
+{
+    if (!hasHistoryActionActivation())
+        return false;
+
+    for (RefPtr<Frame> frame = this->frame() ? &this->frame()->tree().top() : nullptr; frame; frame = frame->tree().traverseNext()) {
```

Comment:
Do we really want to traverse ALL frames?  Does this mean that an iframe will affect its parent frame, even if they are from different origins?

Reply:
This implementation matches that of consumeTransientActivation.

https://html.spec.whatwg.org/multipage/interaction.html#consume-history-action-user-activation shows that the two algorithms are almost identical.

Reply:
That may be true, but it is implemented in a way that is only reachable by test code.  That algorithm is intended to be useable by the web platform.

Reply:
Apologies I don't understand what the issue with them is? 

https://github.com/WebKit/WebKit/blob/8974c3b7c78a830bf62cdb2a5b8384f73e35a696/Source/WebCore/html/closewatcher/CloseWatcher.cpp#L105 here's the line in a follow up patch where I use these

Reply:
Oh, the comment "That may be true..." was meant for the other thread about only being reachable from test code.  I see your plan a bit better now.